encodings/internal.js

```diff
@@ -204,22 +204,33 @@ InternalDecoderCesu8.prototype.end = function() {
 
 function InternalEncoderUtf8(options, codec) {
     this.lowSurrogate = '';
+    this.surrogatePair = '';
 }
 
 InternalEncoderUtf8.prototype.write = function(str) {
-    if (str.length === 1 && this.lowSurrogate && '\uDC00' < str && str <= '\uDFFF') {
-        // is Surrogate Pair
-        var buf = Buffer.from(this.lowSurrogate + str)
-        this.lowSurrogate = '';
-        return buf
-    }
-    if (str.length === 1 && '\uD800' < str && str <= '\uDBFF') {
-        // is Low-Surrogate Code Unit
-        this.lowSurrogate = str
-    } else {
-        this.lowSurrogate = '';
-        return Buffer.from(str, this.enc);
+    var normalStr = '';
+    for (var i = 0; i < str.length; i++) {
```

Comment:
a loop over the whole string is pretty inefficient. Moreover, you don't need to touch internal surrogates - Node processes them correctly already. We only need to adjust boundaries, as these are what Node can't adjust as it doesn't know about them.